### PR TITLE
gh-77465: Increase test coverage for `numbers`

### PR DESCRIPTION
**Pull request by aisk**
Most classes in `numbers` are ABCs, but they have default implementations for some methods, for example, `Complex.__bool__`. Although the impementations are simple, but they should have been tested.

* Issue: gh-77465

### CODE REVIEW

**Review by serhiy-storchaka — COMMENTED** [review]
There is too many code for implementing methods that are not used, just to allow instantiation of the subclass of an abstract class. I suggest to generate them.

```python
import abc

def not_implemented(*args, **kwargs):
    raise NotImplementedError

def concretize(cls):
    for name in dir(cls):
        try:
            value = getattr(cls, name)
            if value.__isabstractmethod__:
                setattr(cls, name, not_implemented)
        except AttributeError:
            pass
    abc.update_abstractmethods(cls)
    return cls
```

You can just apply this decorator to subclasses and remove all dummy methods.

**Comment by aisk** [comment]
Thanks for the review, applied!

File: Lib/test/test_abstract_numbers.py [hunk 1]
```diff
@@ -40,5 +60,136 @@ def test_complex(self):
...
+            @property
+            def imag(self):
+                return self.i
+
+            def __add__(self, other):
+                if isinstance(other, Complex):
+                    return MyComplex(self.imag + other.imag,
+                                     self.real + other.real)
+                if isinstance(other, Number):
+                    return MyComplex(self.imag + 0, self.real + other.real)
```
Comment (serhiy-storchaka) [inline Lib/test/test_abstract_numbers.py]: Could not it be simplified?
```suggestion
                if isinstance(other, Number):
                    return MyComplex(self.imag + other.imag,
                                     self.real + other.real)
                raise NotImplementedError
```

File: Lib/test/test_abstract_numbers.py [hunk 2]
```diff
@@ -40,5 +60,136 @@ def test_complex(self):
...
+                return other % self.n
+
+        # test __divmod__
+        self.assertEqual(divmod(MyReal(3), 2), (1, 1))
+
+        # test __rdivmod__
+        self.assertEqual(divmod(3, MyReal(2)), (1, 1))
+
+        # test __complex__
+        self.assertEqual(complex(MyReal(1)).imag, 0j)
```
Comment (serhiy-storchaka) [inline Lib/test/test_abstract_numbers.py]: Why not test the full value?
```suggestion
        self.assertEqual(complex(MyReal(1)), 1+0j)
```

File: Lib/test/test_abstract_numbers.py [hunk 3]
```diff
@@ -40,5 +60,136 @@ def test_complex(self):
...
+            @property
+            def numerator(self):
+                return self.n
+
+            @property
+            def denominator(self):
+                return self.d
+
+        # test__float__
+        self.assertEqual(float(MyRational(9, 3)), 3)
```
Comment (serhiy-storchaka) [inline Lib/test/test_abstract_numbers.py]: It is better to test with non-integer result (it uses true division).
```suggestion
        self.assertEqual(float(MyRational(5, 2)), 2.5)
```

File: Lib/test/test_abstract_numbers.py [hunk 4]
```diff
@@ -1,14 +1,32 @@
...
+            pass
+    abc.update_abstractmethods(cls)
+    return cls
+
 
 class TestNumbers(unittest.TestCase):
     def test_int(self):
         self.assertTrue(issubclass(int, Integral))
         self.assertTrue(issubclass(int, Complex))
+        self.assertTrue(issubclass(int, Number))
```
Comment (serhiy-storchaka) [inline Lib/test/test_abstract_numbers.py:31]: Test also with Rationale and Real. Test all concrete types with all 5 abstract types.

File: Lib/test/test_abstract_numbers.py (hunk 1 above)
Comment (serhiy-storchaka) [reply Lib/test/test_abstract_numbers.py]: On other hand, `real` and `imag` are only defined for `Complex`, so it should be
```suggestion
                if isinstance(other, Complex):
                    return MyComplex(self.imag + other.imag,
                                     self.real + other.real)
                raise NotImplementedError
```

File: Lib/test/test_abstract_numbers.py (hunk 4 above)
Comment (serhiy-storchaka) [reply Lib/test/test_abstract_numbers.py:31]: Please add also similar checks for other concrete types (test_float, test_complex).

Comment (aisk) [reply Lib/test/test_abstract_numbers.py:31]: Sorry, I forgot to check these, updated!

File: Lib/test/test_abstract_numbers.py [hunk 5]
```diff
@@ -38,8 +38,10 @@ def test_int(self):
         self.assertEqual(1, int(7).denominator)
 
     def test_float(self):
+        self.assertTrue(issubclass(int, Integral))
```
Comment (serhiy-storchaka) [inline Lib/test/test_abstract_numbers.py]: It should check float.

File: Lib/test/test_abstract_numbers.py [hunk 6]
```diff
@@ -48,6 +50,8 @@ def test_float(self):
         self.assertEqual(-7.3, float(-7.3).conjugate())
 
     def test_complex(self):
+        self.assertTrue(issubclass(int, Integral))
```
Comment (serhiy-storchaka) [inline Lib/test/test_abstract_numbers.py]: And this should check complex.

File: Lib/test/test_abstract_numbers.py (hunk 5 above)
Comment (aisk) [reply Lib/test/test_abstract_numbers.py]: 😂 My bad

**Review by serhiy-storchaka — APPROVED** [review]
LGTM.

**Pull request merged**